File: crates/oharness-loop/examples/replay_trajectory.rs
//! `replay_trajectory` — run an agent, then replay the recorded
//! trajectory against [`ReplayLlm`] and verify the re-drive matches
//! the original.
//!
//! This is the "scientific" path: you get bit-for-bit reproducibility
//! of a recorded run without needing the underlying provider's API
//! key, network, or dollars. Replay is how debuggers / post-mortems
//! / paper-supplement reproductions work.
//!
//! `ReplayMode::Positional` (used here) pairs the Nth live
//! `llm.request` in the replay loop with the Nth recorded
//! `llm.response` — no byte-for-byte input comparison, so the replay
//! tolerates minor non-determinism in request shape. `ReplayMode::Strict`
//! adds canonical-JSON equality and emits `critic.failed` on drift,
//! controlled via `DriftPolicy::WarnAndContinue` (default) or
//! `DriftPolicy::Fail`.
//!
//! The example also demonstrates writing the captured trajectory to a
//! JSONL file alongside the in-memory round-trip — that's the
//! on-disk format `oharness_trace::jsonl::read_events` + external
//! tools (`jq`, paper-supplement analysis scripts) consume.
//!
//! Run with:
//!
//! ```bash
//! cargo run --example replay_trajectory -p oharness-loop
//! ```

use async_trait::async_trait;
use oharness_core::{
    CompletionRequest, CompletionResponse, Content, LlmCapabilities, Message, ModelId, StopReason,
    Task, Termination, Usage,
};
use oharness_llm::{ChunkStream, Llm, LlmError};
use oharness_loop::{Agent, ReactLoop};
use oharness_tools::fs::FsToolSet;
use oharness_trace::{DriftPolicy, InMemorySink, ReplayLlm, ReplayMode};
use serde_json::json;
use std::io::Write;
use std::sync::atomic::{AtomicU32, Ordering};
use std::sync::Arc;

struct ScriptedLlm {
    responses: Vec<CompletionResponse>,
    cursor: AtomicU32,
}

#[async_trait]
impl Llm for ScriptedLlm {
    fn name(&self) -> &str {
        "scripted"
    }

    fn capabilities(&self) -> LlmCapabilities {
        LlmCapabilities::default()
    }

    async fn complete(&self, _req: CompletionRequest) -> Result<CompletionResponse, LlmError> {
        let idx = self.cursor.fetch_add(1, Ordering::SeqCst) as usize;
        self.responses
            .get(idx)
            .cloned()
            .ok_or(LlmError::Unsupported("script exhausted"))
    }

    async fn stream(&self, _req: CompletionRequest) -> Result<ChunkStream, LlmError> {
        Err(LlmError::Unsupported("stream"))
    }
}

fn script() -> Vec<CompletionResponse> {
    vec![
        CompletionResponse {
            id: "msg_001".into(),
            model: ModelId::new("scripted-replay-example"),
            content: vec![
                Content::text("Let me look."),
                Content::ToolUse {
                    id: "tu_1".into(),
                    name: "fs_list".into(),
                    input: json!({"path": "."}),
                },
            ],
            stop_reason: StopReason::ToolUse,
            usage: Usage {
                tokens_input: 10,
                tokens_output: 5,
                ..Default::default()
            },
        },
        CompletionResponse {
            id: "msg_002".into(),
            model: ModelId::new("scripted-replay-example"),
            content: vec![Content::text("Found a crates/ directory.")],
            stop_reason: StopReason::EndTurn,
            usage: Usage {
                tokens_input: 20,
                tokens_output: 6,
                ..Default::default()
            },
        },
    ]
}

#[tokio::main]
async fn main() -> Result<(), Box<dyn std::error::Error>> {
    // -----------------------------------------------------------------
    // Phase 1: live run that captures every event into an InMemorySink.
    // -----------------------------------------------------------------
    println!("[phase 1] live run");
    let sink = Arc::new(InMemorySink::new());
    let live_agent = Agent::builder()
        .with_llm(Arc::new(ScriptedLlm {
            responses: script(),
            cursor: AtomicU32::new(0),
        }))
        .with_tools(Arc::new(FsToolSet::new()))
        .with_event_sink(sink.clone())
        .with_loop(Box::new(ReactLoop::new()))
        .with_max_turns(5)
        .build()?;
    let live_outcome = live_agent.run(Task::new("look around")).await?;
    let events = sink.events();
    println!(
        "  termination: {:?}, turns: {}, events: {}",
        live_outcome.termination,
        live_outcome.usage.turns,
        events.len(),
    );

    // -----------------------------------------------------------------
    // (Optional) write the captured trajectory to a JSONL file, the
    // on-disk format external tooling consumes. Uses plain std::io —
    // `FileSink` is the production-grade sink but requires careful
    // drop semantics to flush, so we keep the example self-contained
    // by serializing the captured events directly.
    // -----------------------------------------------------------------
    let tempdir = tempfile::tempdir()?;
    let path = tempdir.path().join("trajectory.jsonl");
    {
        let mut f = std::fs::File::create(&path)?;
        for event in &events {
            let line = serde_json::to_string(event)?;
            writeln!(f, "{line}")?;
        }
    }
    println!(
        "[phase 1.5] wrote {} events → {}",
        events.len(),
        path.display()
    );

    // -----------------------------------------------------------------
    // Phase 2: replay directly from the captured events (same result
    // as reading back from disk via `ReplayLlm::from_path`).
    // -----------------------------------------------------------------
    println!("[phase 2] replay");
    let replay = ReplayLlm::from_events(events, ReplayMode::Positional, DriftPolicy::default())?;
    let replay_agent = Agent::builder()
        .with_llm(Arc::new(replay))
        .with_tools(Arc::new(FsToolSet::new()))
        .with_loop(Box::new(ReactLoop::new()))
        .with_max_turns(5)
        .build()?;
    let replay_outcome = replay_agent.run(Task::new("look around")).await?;

    println!(
        "  termination: {:?}, turns: {}, final: {}",
        replay_outcome.termination,
        replay_outcome.usage.turns,
        last_assistant_text(&replay_outcome.final_messages).unwrap_or_else(|| "<none>".into()),
    );

    // -----------------------------------------------------------------
    // Phase 3: assert the replay matches the live run.
    // -----------------------------------------------------------------
    assert_eq!(replay_outcome.usage.turns, live_outcome.usage.turns);
    assert_eq!(
        replay_outcome.usage.tool_calls,
        live_outcome.usage.tool_calls
    );
    assert!(matches!(
        (&live_outcome.termination, &replay_outcome.termination),
        (Termination::Completed { .. }, Termination::Completed { .. }),
    ));
    assert_eq!(
        last_assistant_text(&live_outcome.final_messages),
        last_assistant_text(&replay_outcome.final_messages),
    );
    println!("[phase 3] replay output matches live run ✔");

    Ok(())
}

fn last_assistant_text(messages: &[Message]) -> Option<String> {
    messages.iter().rev().find_map(|m| match m {
        Message::Assistant { content, .. } => content.iter().find_map(|c| match c {
            Content::Text { text } => Some(text.clone()),
            _ => None,
        }),
        _ => None,
    })
}

File: python-examples/replay_trajectory.py
"""replay_trajectory — record a JSONL trajectory, then replay it
against `ReplayLlm`.

This is the paper-supplement reproducibility path: a recorded
trajectory re-drives an agent with byte-for-byte fidelity, no
provider API key required.

`ReplayMode.positional` (used here) pairs the Nth live
`llm.request` with the Nth recorded `llm.response`. Use
`"strict"` + `drift="fail"` for canonical-JSON equality checks.
"""

import itertools
import json
import tempfile
from pathlib import Path

import oharness


class ScriptedLlm:
    """Two-turn script: tool call, then summary."""

    def __init__(self) -> None:
        self._responses = iter([
            {
                "id": "msg_001",
                "model": "scripted-replay",
                "content": [
                    {"type": "text", "text": "Let me look."},
                    {
                        "type": "tool_use",
                        "id": "tu_1",
                        "name": "fs_list",
                        "input": {"path": "."},
                    },
                ],
                "stop_reason": {"kind": "tool_use"},
                "usage": {"tokens_input": 10, "tokens_output": 5},
            },
            {
                "id": "msg_002",
                "model": "scripted-replay",
                "content": [{"type": "text", "text": "Found a crates/ directory."}],
                "stop_reason": {"kind": "end_turn"},
                "usage": {"tokens_input": 20, "tokens_output": 6},
            },
        ])

    def complete(self, req_json: str) -> str:
        return json.dumps(next(self._responses))


def last_assistant_text(messages):
    for msg in reversed(messages):
        if msg.get("role") != "assistant":
            continue
        for block in msg.get("content", []):
            if block.get("type") == "text":
                return block["text"]
    return None


def main() -> None:
    with tempfile.TemporaryDirectory() as td:
        path = Path(td) / "trajectory.jsonl"

        # --- phase 1: record a live run into a FileSink.
        print(f"[phase 1] live run → {path}")
        sink = oharness.FileSink(str(path))
        live_agent = (
            oharness.Agent.builder()
            .with_llm(oharness.PyLlm(ScriptedLlm(), name="scripted"))
            .with_tools(oharness.FsToolSet())
            .with_event_sink(sink)
            .with_loop(oharness.ReactLoop())
            .with_max_turns(5)
            .build()
        )
        live = json.loads(live_agent.run(oharness.Task("look around")))
        # Drain the writer task before reading the file back.
        sink.flush()
        print(
            f"  termination: {live['termination']['kind']}, "
            f"turns: {live['usage']['turns']}"
        )

        # --- phase 2: replay from the recorded trajectory file.
        print("[phase 2] replay")
        replay = oharness.ReplayLlm.from_path(
            str(path), mode="positional", drift="warn_and_continue"
        )
        replay_agent = (
            oharness.Agent.builder()
            .with_llm(replay)
            .with_tools(oharness.FsToolSet())
            .with_loop(oharness.ReactLoop())
            .with_max_turns(5)
            .build()
        )
        replay_out = json.loads(replay_agent.run(oharness.Task("look around")))
        print(
            f"  termination: {replay_out['termination']['kind']}, "
            f"turns: {replay_out['usage']['turns']}, "
            f"final: {last_assistant_text(replay_out['final_messages'])!r}"
        )

        # --- phase 3: assert match.
        assert live["usage"]["turns"] == replay_out["usage"]["turns"]
        assert live["usage"]["tool_calls"] == replay_out["usage"]["tool_calls"]
        assert live["termination"]["kind"] == replay_out["termination"]["kind"] == "completed"
        assert (
            last_assistant_text(live["final_messages"])
            == last_assistant_text(replay_out["final_messages"])
        )
        print("[phase 3] replay output matches live run ✔")


if __name__ == "__main__":
    main()
